datasetinsights/estimators/siam_mask.py

```
@@ -0,0 +1,2215 @@
+                    self.device
+                ),
+                "label_mask": torch.autograd.Variable(input[6]).to(self.device),
+                "label_mask_weight": torch.autograd.Variable(input[7]).to(
+                    self.device
+                ),
+            }
+
+            outputs = self.model(x)
+
+            rpn_cls_loss, rpn_loc_loss, rpn_mask_loss = (
+                torch.mean(outputs["losses"][0]),
+                torch.mean(outputs["losses"][1]),
+                torch.mean(outputs["losses"][2]),
+            )
+            mask_iou_mean, mask_iou_at_5, mask_iou_at_7 = (
+                torch.mean(outputs["accuracy"][0]),
+                torch.mean(outputs["accuracy"][1]),
+                torch.mean(outputs["accuracy"][2]),
+            )
+
+            cls_weight, reg_weight, mask_weight = (
+                config["loss_weights"]["feature"],
+                config["loss_weights"]["rpn"],
+                config["loss_weights"]["mask"],
+            )
+
+            loss = (
+                rpn_cls_loss * cls_weight
+                + rpn_loc_loss * reg_weight
+                + rpn_mask_loss * mask_weight
+            )
+
+            optimizer.zero_grad()
+            loss.backward()
+
+            # Clip gradients of all 3 losses
+            if config["clip"]:
+                torch.nn.utils.clip_grad_norm_(
+                    model.module.features.parameters(),
+                    config["clip"]["feature"],
+                )
+                torch.nn.utils.clip_grad_norm_(
+                    model.module.rpn_model.parameters(), config["clip"]["rpn"]
+                )
+                torch.nn.utils.clip_grad_norm_(
+                    model.module.mask_model.parameters(), config["clip"]["mask"]
+                )
+
+            def is_valid_number(x):
+                return not (math.isnan(x) or math.isinf(x) or x > 1e4)
+
+            if is_valid_number(loss.item()):
+                optimizer.step()
```

when would the loss be invalid? it seems odd to silently fail and keep going.